a10_octavia/common/utils.py

```diff
@@ -91,23 +91,36 @@ def convert_to_hardware_thunder_conf(hardware_list):
     hardware_dict = {}
     for hardware_device in hardware_list:
         hardware_device = validate_params(hardware_device)
-        project_id = hardware_device['project_id']
-        if hardware_dict.get(project_id):
-            raise cfg.ConfigFileValueError('Supplied duplicate project_id {} '
-                                           ' in [hardware_thunder] section'.format(project_id))
         hardware_device['undercloud'] = True
         if hardware_device.get('interface_vlan_map'):
             hardware_device['device_network_map'] = validate_interface_vlan_map(hardware_device)
             del hardware_device['interface_vlan_map']
-        hierarchical_mt = hardware_device.get('hierarchical_multitenancy')
-        if hierarchical_mt == "enable":
-            hardware_device["partition_name"] = project_id[0:14]
-        if hierarchical_mt and hierarchical_mt not in ('enable', 'disable'):
-            raise cfg.ConfigFileValueError('Option `hierarchical_multitenancy` specified '
-                                           'under project id {} only accepts "enable" and '
-                                           '"disable"'.format(project_id))
+
+        # Add dict entries with project_id as key
+        if 'project_id' in hardware_device:
+            project_id = hardware_device['project_id']
+            if hardware_dict.get(project_id):
+                raise cfg.ConfigFileValueError('Supplied duplicate project_id {} '
+                                               ' in [hardware_thunder] section'.format(project_id))
+            hierarchical_mt = hardware_device.get('hierarchical_multitenancy')
+            if hierarchical_mt == "enable":
+                hardware_device["partition_name"] = project_id[0:14]
+            if hierarchical_mt and hierarchical_mt not in ('enable', 'disable'):
+                raise cfg.ConfigFileValueError('Option `hierarchical_multitenancy` specified '
+                                               'under project id {} only accepts "enable" and '
+                                               '"disable"'.format(project_id))
+            vthunder_conf = data_models.HardwareThunder(**hardware_device)
+            hardware_dict[project_id] = vthunder_conf
+
+        # Add dict entries with dev_name as key
+        device_name = hardware_device['device_name']
+        if hardware_dict.get(a10constants.DEVICE_KEY_PREFIX + device_name):
+            raise cfg.ConfigFileValueError('Supplied duplicate device_name {} '
+                                           ' in [hardware_thunder] section'.format(device_name))
         vthunder_conf = data_models.HardwareThunder(**hardware_device)
-        hardware_dict[project_id] = vthunder_conf
+        vthunder_conf.device_name_as_key = True
```

Comment:
Maybe indentation is weird on github.  Looks like we are always setting `device_name_as_key` as true. Which means the check duplicate logic for HMT is never called.

Reply:
To follow onto this, we don't know that they want to use the device name for creation until they actively use the flavor.

Reply:
Missed the iteration during review. nvm